[torchgen] Generate out variant for functional operator

larryliu0820 opened the pull request:
Summary:
Previously we don't generate out variant (both schema and kernel) for an operator with functional variant only. This adds support for that and adds test.


## Changes on `native_function_generation.py`

We are generating out variant for all functional variants if possible. This PR introduces a lot of newly generated out variants and `native_functions.yaml` needs to incorporate the changes by adding `autogen` keywords. 

The logic for determining what operators we should generate an out variant for is the following:

1. No existing out variant for this `NativeFunction`
2. Contains an existing in place, mutable or functional variant
3. Contains at least 1 tensor like return(s)

For operators matching the first two conditions but failing the third, I listed them in `FUNCTIONAL_OPS_THAT_CANNOT_GET_AN_OUT_VARIANT`.

## Special handling

The following operators satisfy all 3 criteria above but we chose to not autogen them, with some reasons.
* `mkldnn_adaptive_avg_pool2d`, the generated out variant `mkldnn_adaptive_avg_pool2d.out` is colliding with the `mkldnn_adaptive_avg_pool2d_out` kernel in `adaptive_avg_pool2d.out` operator. I manually created `mkldnn_adaptive_avg_pool2d.out` and renamed `mkldnn_adaptive_avg_pool2d_out` to `mkldnn_adaptive_avg_pool2d_out_stub`.
* `min`, `max` and `mean`. There already exist `min.out`, `max.out` and `mean.out` but they are having different semantics with the functional ones. I manually created `min.unary_out`, `max.unary_out` and `mean.dtype_out` to disambiguate.


## Autograd Changes

We introduced a logic to not match derivatives info in `derivatives.yaml` to out variant, since we are generating `NOT_IMPLEMENTED` kernels for those out variants anyway. The issue we are seeing with the original logic is that it doesn't handle `TensorOption` arguments really well. For example we have these two operators:

* `_to_copy(Tensor self, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None, bool non_blocking=False, MemoryFormat? memory_format=None) -> Tensor`
* `_to_copy.out(Tensor self, *, bool non_blocking=False, MemoryFormat? memory_format=None, Tensor(a!) out) -> Tensor(a!)`

If we uses `_to_copy` derivative info, there will be compilation error since `dtype` is missing from `_to_copy.out` signature.
Test Plan: Rely on unit test

Differential Revision: D37832342

bdhirsh commented on torchgen/model.py:
```diff
@@ -1240,8 +1242,8 @@ def __post_init__(self) -> None:
-            assert self.kind() == SchemaKind.functional, (
-                "Found an operator that is not functional, but has tensor options arguments."
+            assert self.kind() == SchemaKind.functional or self.kind() == SchemaKind.out, (
+                "Found an operator that is not functional or out varuabt, but has tensor options arguments."
```
This change seems suspicious - we don't have / want any out= ops to take in `TensorOptions` arguments.

We technically do have some factory functions that have out= variants ([example](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/native_functions.yaml#L628)), but you can see that they don't take in `TensorOptions` arguments - the metadata can be inferred from the `out=` tensor.

bdhirsh replied:
bump :)

bdhirsh commented on torchgen/native_function_generation.py:
```diff
@@ -103,27 +105,32 @@ def self_to_out_signature(func: FunctionSchema) -> FunctionSchema:
+# Helper function: given a functional FunctionSchema, generate its corresponding out= variant
 # Example before:
-#   _fused_moving_avg_obs_fq_helper(Tensor self, Tensor observer_on, Tensor fake_quant_on, Tensor(a!) running_min, Tensor(b!) running_max, Tensor(c!) scale, Tensor(d!) zero_point, float averaging_const, int quant_min, int quant_max, int ch_axis, bool per_row_fake_quant=False, bool symmetric_quant=False) -> (Tensor output, Tensor mask)  # noqa: B950
+#   _to_copy(Tensor self, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None, bool non_blocking=False, MemoryFormat? memory_format=None) -> Tensor
```
ah ok - I think we should it looks like `_to_copy()` takes in the full set of TensorOptions arguments.

I think that to fit the pattern that we have for other factory functions with out= variants, we should arrange for the signature to look like:
```
_to_copy.out(Tensor self, *, bool non_blocking=False, MemoryFormat? memory_format=None, Tensor(a!) out) -> Tensor(a!)
```

larryliu0820 replied:
ok, if we set `_to_copy.out` signature like that, `autogen` won't work on it?

bdhirsh commented:
@larryliu0820 lmk when you want me to take another look! IIRC, what we agreed on last week was:
- we should probably just add all of the functional ops at once instead of adding a mechanism to add one at a time
- This PR should only bother doing it for non-composite ops, which is the main blocker for getting functionalization working. So there shouldn't be **too** many that you need to manually add the the yaml (since 90% of the missing out= variants are coming from composites).

iseeyuan commented:
> 

Thanks. I agree with both of the points.

bdhirsh commented on torchgen/native_function_generation.py:
```diff
@@ -103,27 +131,42 @@ def self_to_out_signature(func: FunctionSchema) -> FunctionSchema:
-    #     but if necessary we could probably add another out= variant that also
-    #     functionalizes the mutable arguments (a functional_out variant)
+    arguments: Arguments = (
+        func.arguments.remove_self_annotation()
```
I guess view ops currently show up as having a functional schema. But we should probably just skip view ops here, and not generate out= variants for them in core.

bdhirsh commented on torchgen/native_function_generation.py:
```diff
@@ -103,27 +131,42 @@ def self_to_out_signature(func: FunctionSchema) -> FunctionSchema:
-    # - Any non-aliased tensor-like returns are converted to mutable, aliased out= arguments
-    #   (if the argument is a tensor then we also return it for method chaining,
-    #   otherwise we return nothing)
+    # - a new out argument with the same type as "func" (but with a mutable annotation)
```
What's "func" here? (maybe this is referring to the return argument in the functional op)

bdhirsh commented on torchgen/native_function_generation.py:
```diff
@@ -218,15 +286,17 @@ def generate_function(
             "We currently only generate either functional or out= NativeFunctions"
         )
-
+    # generated kernel has a naming convention `.*_generated` to avoid collision with existing operator name.
```
where does this happen?

bdhirsh commented on torchgen/native_function_generation.py:
```diff
@@ -359,14 +428,17 @@ def add_generated_native_functions(
+                    not in FUNCTIONAL_OPS_THAT_CANNOT_GET_AN_OUT_VARIANT
                 ):
                     raise AssertionError(
                         f"""Found a mutable operator that we could not generate an out= variant for: {str(base_fn.func)}.
```
Nit: "Found a mutable operator" -> "Found an operator", now that the op we're generating out= variants of can be functional too.

It's probably worth updating this error message a bit too: if someone adds a new functional op that returns a non-tensor output, the error message should probably tell them to directly add their op name to the `FUNCTIONAL_OPS_THAT_CANNOT_GET_AN_OUT_VARIANT` list.

bdhirsh commented on torchgen/api/autograd.py:
```diff
@@ -312,6 +312,10 @@ def match_differentiability_info(
 
     def find_info(f: NativeFunction) -> Tuple[Optional[DifferentiabilityInfo], bool]:
+        # Don't bother matching info to generated out= variants
+        if f.func.kind() == SchemaKind.out:
```
Hmm, should this be `if "generated" in f.tags and f.func.kind() == SchemaKind.out`? We probably don't need to bother routing the old out= ops through this logic, just the new generated ones.

bdhirsh commented on tools/test/test_codegen.py:
```diff
@@ -270,6 +276,35 @@ def test_native_function_declaration_1_op_1_ns_valid(self) -> None:
+
+    def test_functional_variant_autogen_out_variant(self) -> None:
+        add_generated_native_functions(self.native_functions, self.backend_indices)
+        self.assertEqual(len(self.native_functions), 2)
```
it might be useful for this to test an edge case: e.g. if the schema of `op` were `op(Tensor self) -> (Tensor, Tensor)`, then we'd expect the out= argument of the generated function to be a `Tuple[Tensor, Tensor]`.

bdhirsh commented:
github isn't even letting me load the `native_functions.yaml` diff 😛

larryliu0820 committed:
[torchgen] Generate out variant for functional operator (#81437)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/81437

Previously we don't generate out variant (both schema and kernel) for an operator with functional variant only. This adds support for that and adds test.

We are generating out variant for all functional variants if possible. This PR introduces a lot of newly generated out variants and `native_functions.yaml` needs to incorporate the changes by adding `autogen` keywords.

The logic for determining what operators we should generate an out variant for is the following:

1. No existing out variant for this `NativeFunction`
2. Contains an existing in place, mutable or functional variant
3. Contains at least 1 tensor like return(s)

For operators matching the first two conditions but failing the third, I listed them in `FUNCTIONAL_OPS_THAT_CANNOT_GET_AN_OUT_VARIANT`.

The following operators satisfy all 3 criteria above but we chose to not autogen them, with some reasons.
* `mkldnn_adaptive_avg_pool2d`, the generated out variant `mkldnn_adaptive_avg_pool2d.out` is colliding with the `mkldnn_adaptive_avg_pool2d_out` kernel in `adaptive_avg_pool2d.out` operator. I manually created `mkldnn_adaptive_avg_pool2d.out` and renamed `mkldnn_adaptive_avg_pool2d_out` to `mkldnn_adaptive_avg_pool2d_out_stub`.
* `min`, `max` and `mean`. There already exist `min.out`, `max.out` and `mean.out` but they are having different semantics with the functional ones. I manually created `min.unary_out`, `max.unary_out` and `mean.dtype_out` to disambiguate.

We introduced a logic to not match derivatives info in `derivatives.yaml` to out variant, since we are generating `NOT_IMPLEMENTED` kernels for those out variants anyway. The issue we are seeing with the original logic is that it doesn't handle `TensorOption` arguments really well. For example we have these two operators:

* `_to_copy(Tensor self, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None, bool non_blocking=False, MemoryFormat? memory_format=None) -> Tensor`
* `_to_copy.out(Tensor self, *, bool non_blocking=False, MemoryFormat? memory_format=None, Tensor(a!) out) -> Tensor(a!)`

If we uses `_to_copy` derivative info, there will be compilation error since `dtype` is missing from `_to_copy.out` signature.

Test Plan: Rely on CI and unit test

Differential Revision: D37832342

fbshipit-source-id: bb568eee46e2723634997400b9057fbed162dce1

bdhirsh reviewed (approved):
other than the nits we talked about, looks good to me! LMK if you need help with the OSS failures

larryliu0820 commented:
@pytorchbot merge
